feat: Allow multi party reconciliation against a Bank Transaction

marination opened the pull request:
- Use a Journal Entry to create a party agnostic settlement
- ui: Popup to warn user and ask for JE account in case of multi party selection. The **Debit To**, **Credit To** or **Payable Account** is fetched by the backend depending on the kind of vouchers selected
- ![2024-10-10 12 31 35 PM](https://github.com/user-attachments/assets/7794847d-8494-450d-b739-d1877f647c69)
-  <img width="600" alt="Screenshot 2024-10-09 at 6 47 59 PM" src="https://github.com/user-attachments/assets/c98d3234-582f-4b8d-9253-0612049d31cd">
-  UI Validation (as well as backend, that already existed) against selecting vouchers of different types
-  <img width="700" alt="Screenshot 2024-10-10 at 12 38 19 PM" src="https://github.com/user-attachments/assets/461ec1c2-065b-4727-a008-42f53c858009">


**ToDo:**

- [x] Code cleanup and abstractions
- [x] Adjust tests

marination committed:
feat: Allow multi party reconciliation against a Bank Transaction

- Use a Journal Entry to create a party agnostic settlement
- ui: Popup to warn user and ask for JE account in case of multi party selection

marination committed:
test: Multi Party Reconciliation

- misc: Use `bulk_reconcile_vouchers` instead of erpnext's `reconcile_vouchers`

marination commented:
@barredterra Do we need a setting here since the popup already gives the user a heads up. If someone does not desire this and it was a mistake they should just shut the dialog. Or do we want to make sure this not accessible if a company does not need it ?

barredterra commented:
@marination I didn't look at the code yet, but from the screenshots I wondered: shouldn't we use the voucher's _Debit To_ / _Credit To_ accounts, instead of letting the user set one manually?

marination committed:
fix: Use debit to/credit to account from the invoice

- Do not ask for account from the popup, just use prompt to warn user
- Fetch the second account from the first invoice. Assumption: All invoices are of the same kind (validation exists) and use the same intermediate account
- Remove `account` arg that was drilled down

marination committed:
chore: Common function for repetitive logic: `adjust_and_allocate_invoices`

- misc: Extract `get_debtor_creditor_account` into external utility.It does not need the class attributes
- `adjust_and_allocate_invoices` attaches invoices with the right amounts to the payemt voucher. It avoids over-allocation
- It also accepts a callable that executes at the end of the allocation

marination committed:
fix: Get intermediate account for each invoice row

- Different Sales Invoices could have different Debit To accounts. These should reflect in the Journal Entry on each row

marination committed:
chore: Translations

barredterra committed:
test: extend test case for mulri_party reco

barredterra commented on banking/overrides/bank_transaction.py:
```diff
@@ -282,6 +385,20 @@ def get_outstanding_amount(payment_doctype, payment_name) -> float:
 
+def get_debtor_creditor_account(invoice: dict) -> str | None:
+	"""Get the debtor or creditor (intermediate) account based on the invoice type."""
+	account_field = (
+		"debit_to"
+		if invoice.get("voucher_type") == "Sales Invoice"
+		else "credit_to"
+		if invoice.get("voucher_type") == "Purchase Invoice"
+		else "payable_account"
+	)
```
```suggestion
		if invoice.get("voucher_type") == "Sales Invoice":
			account_field = "debit_to"
		elif invoice.get("voucher_type") == "Purchase Invoice":
			account_field = "credit_to"
		else:
			account_field = "payable_account"
```

barredterra commented on banking/public/js/bank_reconciliation_beta/actions_panel/match_tab.js:
```diff
@@ -253,6 +275,25 @@ erpnext.accounts.bank_reconciliation.MatchTab = class MatchTab {
 	}
 
+	show_multiple_party_reconcile_prompt(selected_vouchers) {
+		let me = this;
+		let fields = [
+			{
+				fieldtype: "HTML",
+				fieldname: "info",
+				options: __("Are you trying to reconcile vouchers of different parties? This action will reconcile vouchers using a Journal Entry."),
+			},
+		]
+		frappe.prompt(
+			fields,
+			() => {
+				me.bulk_reconcile_vouchers(selected_vouchers, true);
+			},
+			__("Multiple Party Reconciliation"),
+			__("Confirm")
+		);
```
```suggestion
		frappe.confirm(
			__("Are you trying to reconcile vouchers of different parties? This action will reconcile vouchers using a Journal Entry."),
			() => this.bulk_reconcile_vouchers(selected_vouchers, true);
		);
```

marination committed:
chore: Improve readability on setting `account_field`

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

marination committed:
chore: Use less verbose `frappe.confirm`

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

barredterra commented:
@Mergifyio backport version-14
